src/aiida/orm/fields.py

```diff
@@ -0,0 +1,499 @@
+        """Return an QbField by key."""
+        return self._fields[key]
+
+    def __getattr__(self, key: str) -> QbField:
+        """Return an QbField by key."""
+        try:
+            return self._fields[key]
+        except KeyError:
+            raise AttributeError(key)
+
+    def __contains__(self, key: str) -> bool:
+        """Return if the field key exists"""
+        return key in self._fields
+
+    def __len__(self) -> int:
+        """Return the number of fields"""
+        return len(self._fields)
+
+    def __iter__(self):
+        """Iterate through the field keys"""
+        return iter(self._fields)
+
+    def __dir__(self):
+        """Return keys for tab competion."""
+        return list(self._fields) + ['_dict']
+
+    @property
+    def _dict(self):
+        """Return a copy of the internal mapping"""
+        return deepcopy(self._fields)
+
+
+class EntityFieldMeta(ABCMeta):
+    """A metaclass for entity fields, which adds a `fields` class attribute."""
+
+    def __init__(cls, name, bases, classdict):
+        super().__init__(name, bases, classdict)
+
+        # only allow an existing fields attribute if has been generated from a subclass
+        current_fields = getattr(cls, 'fields', None)
+        if current_fields is not None and not isinstance(current_fields, QbFields):
+            raise ValueError(f"class '{cls}' already has a `fields` attribute set")
+
+        # Find all fields
+        fields = {}
+        # Note: inspect.getmembers causes loading of AiiDA to fail
+        for key, attr in ((key, attr) for subcls in reversed(cls.__mro__) for key, attr in subcls.__dict__.items()):
+            # __qb_fields__ should be a list of QbField instances
+            if key == '__qb_fields__':
+                assert isinstance(
+                    attr, t.Sequence
+                ), f"class '{cls}' has a '__qb_fields__' attribute, but it is not a sequence"
+                for field in attr:
+                    if not isinstance(field, QbField):
+                        raise ValueError(f"__qb_fields__ attribute of class '{cls}' must be list of QbField instances")
+                    if field.key == 'value':
+                        if dtype := getattr(cls, '_type', None):
+                            value_field = add_field(
+                                'value',
+                                dtype=dtype,
+                                doc='The value of the data',
+                            )
+
+                            fields[field.key] = value_field
+                    else:
+                        fields[field.key] = field
+
+        cls.fields = QbFields({key: fields[key] for key in sorted(fields)})
+
+
+class QbFieldArguments(t.TypedDict):
+    key: str
+    qb_field: t.Optional[str]
+    dtype: t.Optional[t.Any]
+    doc: str
+    is_db_field: bool
+    is_subscriptable: bool
+
+
+def add_field(
+    key: str,
+    qb_field: t.Optional[str] = None,
```

Comment:
Is the name `qb_field` historical? Wonder if this is the most intuitive name. It really is an alias for attributes right? So maybe make it more explicit and name it `attribute_alias`? In the same light, would it maybe also be better to rename `is_db_field` to `is_attribute` (you can set the default to `False` so you keep the plugin writers from not having to explicitly define it). After all, both these keys are just to deal with an exception we have for the `attributes` column

Reply:
You'd have to ask Chris about `qb_field`. I'm open to a change. I was considering `alias` per our chat, but then realized that it's really the frontend `key` that is the alias to the "real" backend key. Thoughts?

`is_db_field` used to be `is_attribute` with the opposite meaning. However, once it was understood that the default behavior of "not an attribute" is ideal w.r.t plugin developers, I decided to switch. So `is_attribute` is not right for non-attribute fields, which are DB columns, hence `db_field`. But maybe `is_column`?

Note that it's not just the new guard that uses these fields. If a field is an attribute (currently not `is_db_field`), the `qb_field` property will return `f'attributes.{self._qb_field}'`.

Reply:
Just remembered, I initially had `is_db_field` still as `is_attribute` but set it to `True` by default to support plugin developers. I had switched due to some annoying printing in `__reprt__` and `__str__`, but I think this can be modified. Let me commit what I have in mind for your review.

Still not sure what to call `qb_field` though due to the confusion about which is really the alias

Reply:
I think I'm fine with the backend key being the alias to the frontend, so `qb_field` -> `alias`. Agreed?

Reply:
> I was considering alias per our chat, but then realized that it's really the frontend key that is the alias to the "real" backend key. Thoughts?

Sure, both points of view can be defended. However, I think the point of this functionality is to provide a single unambiguous interface for users to determine fields of ORM entities, _regardless_ of the backend that is used. In that sense, I think the front-end keys are leading and if the backend decides to do something else, _that_ is the exception. We only need that alias because some existing `Data` plugins made the choice a long time ago to use attribute keys that are not valid Python identifiers. If we were to make the choice now, we probably wouldn't have done this and not even have needed an alias system.

Also, we already concluded that the only place that currently has such an exception are the attributes. That is why I think it makes most sense to have the argument that switches it, be called `is_attribute`.
> However, once it was understood that the default behavior of "not an attribute" is ideal w.r.t plugin developers, I decided to switch.

We can make `is_attribute` default to `True`, so plugin developers don't have to actually use it. However, the implementation is going to make a lot more sense. I don't think `is_column` is a good idea, because it ties it again too much to a particular backend implementation. It is possible to have a storage backend that doesn't use a relational database and so doesn't have the concept of "columns".

Reply:
If `qb_field` goes to `alias`, what should `__qb_fields__` go to? `__aliases__` seems vague. Note that this is likely not important considering the upcoming switch to `pydantic` models.

Reply:
Maybe `alias` is also vague. `backend_alias`?

Reply:
Wait, `__qb_fields__` will hold instances of `QbField` right? Not just strings. Anyway, it is true that it is a bit weird now that it is directly coupled to the querybuilder where it is ultimately not exclusive to the query builder. But as you say, that will be changed when we move to pydantic.

> Maybe alias is also vague. backend_alias?

What is wrong with `attribute_alias`? That is literally what it is right? It crops up only if the field defines an attribute, but the actual key used in the attributes is different (for historical reasons). Since it only ever is going to be related to attributes, I don't see a reason to not make this very explicit in the name

Reply:
I think `alias` is actually sufficient. It is just the name of the parameter to `add_field` and `QbField`. Internally, I'm calling it `backend_key`, which is what it is. In `querybuilder.py`, you then get `projection.backend_key`, which is clear.

Reply:
Ugh, something is breaking. I'll look at it in the morning.

Reply:
Okay @sphuber, all good. Please take a looksee.